caas/kubernetes/provider/bootstrap.go

```diff
@@ -0,0 +1,758 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/juju/errors"
+	"gopkg.in/juju/names.v2"
+	apps "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	k8sstorage "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	intstr "k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/juju/juju/agent"
+	"github.com/juju/juju/cloudconfig/podcfg"
+	"github.com/juju/juju/mongo"
+)
+
+const (
+	// JujuControllerStackName is the juju CAAS controller stack name.
+	JujuControllerStackName = "juju-controller"
+
+	portMongoDB             = 37017
+	portAPIServer           = 17070
+	fileNameSharedSecret    = "shared-secret"
+	fileNameSSLKey          = "server.pem"
+	fileNameBootstrapParams = "bootstrap-params"
+	fileNameAgentConf       = "agent.conf"
+
+	storageSizeControllerRaw = "20Gi" // TODO(caas): parse from constrains?
+)
+
+var (
+	stackLabelsGetter                       = func(stackName string) map[string]string { return map[string]string{labelApplication: stackName} }
```

Comment:
Instead of having all these func types as package scoped vars, we'd have a bootstrapper struct and a NewBootstrapper() method. And for testing, we'd have a NewBootstrapperForTest() and pass in all the fake implementations etc. Like we do for cmd/juju CLI structs.

Reply:
yeah, wrapped all into struct.